Use Symfony and FOSRestBundle to serve languages API

nickygerritsen opened the pull request:
This is one way we could to the API using Symfony.

I've chosen languages on purpose: this is one of the easier API's.

There are still some things to be fixed / figured out:
* Most other API's depend on a contest. We can probably use a (custom) paramconverter to read the contest from the URL and make sure it is active and you have access. I did this in my branch from half a year ago and we can probably reuse some stuff from there
* Some endpoints should return different things based on permissions. We can probably use the expression exclusion strategy for this: https://jmsyst.com/libs/serializer/master/cookbook/exclusion_strategies
* `eventlog` needs to be able to get multiple items at once by passing multiple ID's. The current (v4) implementation uses a comma-separated list of ID's and some ugly magic, but I think it is better to do a POST with a JSON-encoded array or something. Maybe @ubergeek42 has idea's about this?
* Should we indeed do one controller per API 'part'? I think this is best, because later we might want to add more than only GET's.
* How do we handle `ORIDNAL` types using this method? For example, `GET <api>/contests/<id>/problems` needs a numeric position in each element with its index.
* Maybe more a general Doctrine/Symfony thing, but sometimes we only want some columns or only the count of an entity and we do not want to select everything from the table for that. For example the `testcase_count` for problems will, if you just do `->count()` on the property, query and hydrate the whole rows, which should not be done because it is slow. Maybe @ubergeek42 has an idea about this as well?
* We should think of a way to distinguish between returning internal and external ID's with some config option. We can probably use the `Group` feature of JMS Serializer for this.

nickygerritsen commented on webapp/src/DOMJudgeBundle/Controller/API/LanguageController.php:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /**
+     * @Rest\Get("/{externalid}")
```
Currently v4 always uses externalid, but this is one of my comments that we should look into this

ubergeek42 commented on webapp/src/DOMJudgeBundle/Controller/API/LanguageController.php:
```diff
@@ -0,0 +1,48 @@
+    public function getLanguagessAction()
+    {
+        return $this->entityManager->createQueryBuilder()
+            ->from('DOMJudgeBundle:Language', 'l')
```
They can also be multiple characters, so you could do `lang`

nickygerritsen commented:
We might also solve some problems with the JMS Serializer handler support

nickygerritsen commented:
To answer this myself:

> Maybe more a general Doctrine/Symfony thing, but sometimes we only want some columns or only the count of an entity and we do not want to select everything from the table for that. For example the testcase_count for problems will, if you just do ->count() on the property, query and hydrate the whole rows, which should not be done because it is slow. Maybe @ubergeek42 has an idea about this as well?

We can mark associations as extra lazy, and then we can do `->count()` without all being selected.

nickygerritsen commented:
I updated this PR. It now has a nice abstract base controller which does a lot of the shared work. Although we can not use it 100% for all API actions, I think this will be very useful to easily add API actions.

I also added a way to get multiple objects without using a comma-separated list of ID's in GET.

I think I know how to solve most of my questions. I'm wondering how we should continue with this: should we keep it on a separate branch while adding more actions or should we just merge it? It uses a `/v5` prefix for the API, so it will not break anything and by merging we at least make sure we do not diverge to much. Any ideas?

meisterT commented on webapp/src/DOMJudgeBundle/Controller/API/AbstractRestController.php:
```diff
@@ -0,0 +1,139 @@
+    {
+        $ids = $request->request->get('ids', []);
+        if (!is_array($ids) || empty($ids)) {
+            throw new BadRequestHttpException('Please provider a field \'ids\' in the body with an array of ID\'s to fetch');
```
s/provider/provide

meisterT commented:
> I updated this PR. It now has a nice abstract base controller which does a lot of the shared work. Although we can not use it 100% for all API actions, I think this will be very useful to easily add API actions.

That looks indeed like a good idea!


> I also added a way to get multiple objects without using a comma-separated list of ID's in GET.

Disadvantage is that it's harder to call this from the commandline, e.g. via curl. But I agree that it's cleaner.


> I think I know how to solve most of my questions. I'm wondering how we should continue with this: should we keep it on a separate branch while adding more actions or should we just merge it? It uses a `/v5` prefix for the API, so it will not break anything and by merging we at least make sure we do not diverge to much. Any ideas?

I'm definitely for merging. However, I would also go one step further and don't do a v5 for that (aka remove the old API endpoint and replace it with this one). It only changes change implementation and not signature of the API, right?

nickygerritsen commented:
> > I also added a way to get multiple objects without using a comma-separated list of ID's in GET.
> 
> Disadvantage is that it's harder to call this from the commandline, e.g. via curl. But I agree that it's cleaner.
Yeah, although if you use httpie, it is as easy as `http --form --auth admin:admin POST http://localhost:12345/api/v5/contests/3/languages 'ids[]=c' 'ids[]=cpp'`
> 
> > I think I know how to solve most of my questions. I'm wondering how we should continue with this: should we keep it on a separate branch while adding more actions or should we just merge it? It uses a `/v5` prefix for the API, so it will not break anything and by merging we at least make sure we do not diverge to much. Any ideas?
> 
> I'm definitely for merging. However, I would also go one step further and don't do a v5 for that (aka remove the old API endpoint and replace it with this one). It only changes change implementation and not signature of the API, right?
Not entirely, the current API is both at `/v4/` and at `/` I think. `eventlog` uses the `/` one and it will break as soon as we port runs and judgings to it, as that uses multiple ID's. Although we can fix eventlog at that point to do the POST.

I think Travis checks whether the API is correct even right, for the CCS endpoints? So then we can move it to v4 (and non-versioned) and see if Travis is still happy. Then we can do a PR per (couple of) endpoint(s). Does that sound like a good idea?

nickygerritsen commented:
I have done some more things to this branch:

* I have cleaned up the code a bit
* I have declared strict types in the classes (after merging master)
* I have implemented the getting of the contest-endpoint (other calls still TODO)
* I have moved away from `/api/v5` and moved to `/api/v4` as well as just `/api`
* I changed a bit in the abstract controller, making the methods protected, forcing you to create your own public methods. See the last point as to why
* Used the awesome NelmioApiDocBundle to fully automatically generate Swagger API documentation for the API!

This last point means that if you visit `<dj>/api/doc` you get this interface:
<img width="1792" alt="screenshot 2018-09-29 at 20 35 45" src="https://user-images.githubusercontent.com/550145/46249316-4bd7c280-c427-11e8-98af-1a83b92021c2.png">
(it still has a lot of the 'old' endpoints, some of which we definitely need to clean up)

Opening up one of the endpoints gives you something like this:
<img width="1792" alt="screenshot 2018-09-29 at 20 37 08" src="https://user-images.githubusercontent.com/550145/46249322-76c21680-c427-11e8-8e58-4ad93c994e5e.png">

All the models are also automatically shown, so you know what kind of data to expect. You can also authenticate with username/password so calls are done using these credentials.

I see travis is broken on the submit client, so I'll investigate that, but if that gets fixed I'd really like to merge this, so we have a good base for continuing this effort :) .

nickygerritsen commented:
I merged the wrong way around by accident. I have fixed Travis afterwards however.
If anyone disagrees with merging this let me know and I’ll do some git magic

thijskh commented:
Approach looks good to me, probably fine to merge it and solve any issues when found

eldering commented:
I'm sorry I haven't had time yet to take a good look at this, but looks good to me. I had a quick glance at the NelmioApiDocBundle and it looks nice indeed.
Just a nitpick: the POST endpoints to get a list of objects are temporary I assume?

nickygerritsen commented:
@eldering the POST endpoints to get a list of objects replace the current way of getting multiple items at the same time, which is used in `eventlog` during a rejudging to get all runs and judgings at the same time. We currently do this by a GET with all ID's separated by a comma, but that has disadvantages (like GET max length and ID's not being able to have a comma).

I think the POST approach is better. We can also move these endpoints to `POST contests/<cid>/<endpoint>/multiple` or something. Or do you have an other / better idea to solve the eventlog problem?

eldering commented:
> @eldering the POST endpoints to get a list of objects replace the current way of getting multiple items at the same time, which is used in `eventlog` during a rejudging to get all runs and judgings at the same time. We currently do this by a GET with all ID's separated by a comma, but that has disadvantages (like GET max length and ID's not being able to have a comma).
> 
> I think the POST approach is better. We can also move these endpoints to `POST contests/<cid>/<endpoint>/multiple` or something. Or do you have an other / better idea to solve the eventlog problem?

The problem I see with using POST is twofold: first it's against REST principles, since this is not actually creating new records. Secondly, at these particular endpoints, it will be in the way of any real POST operations we have or are going to define. For example to submissions, judgings, judging_runs.

nickygerritsen commented:
The second point is why we should at least move it.

For the first point I don’t really see a solution, as rest only allows GET.
I am thinking if maybe we can rewrite the eventlog function to not use the API completely, but just use the Symfony Doctrine entities directly and Serialize them. What do you think of this idea? This would mean we can drop the POST’s

nickygerritsen commented:
Hmmm after thinking a bit I think skipping the api is hard (and not nice anyway), but what we could do is change the GET’s for the non-id version (i.e. /contests/<id>/languages) to allow a query parameter in the form `ids[]=cpp&ids[]=c`. That would be the cleanest I guess?

eldering commented:
> Hmmm after thinking a bit I think skipping the api is hard (and not nice anyway), but what we could do is change the GET’s for the non-id version (i.e. /contests//languages) to allow a query parameter in the form `ids[]=cpp&ids[]=c`. That would be the cleanest I guess?

That sounds like a good solution to me, although I was hoping that we do an internal bypass of the API. Is `ids[]=cpp&ids[]=c` standard syntax to pass a list as URL parameter?

nickygerritsen commented:
Yeah it is either `ids=cpp&ids=c` or with `[]`. I'll see which one is easier to implement.

We *could* bypass the API, but then we need to move all logic outside of the controllers into some classes that we can put into the Symfony container and expose them. I think this is too much work compared to the benefits.

I'll change the POST's to my proposed solution :)

nickygerritsen commented:
`ids[]=` seems to be the way to go: https://stackoverflow.com/questions/11978438/how-to-get-value-for-multivalued-get-parameter-in-symfony2 (and also `$_GET` seems to do that when passing `ids[]=` and not `ids=`)

nickygerritsen commented:
I have now removed the POST actions and added `ids[]=` support to the new endpoints
